common/archiver/s3store/visibilityArchiver.go

```diff
@@ -0,0 +1,252 @@
+
+package s3store
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/service/config"
+)
+
+type (
+	visibilityArchiver struct {
+		container   *archiver.VisibilityBootstrapContainer
+		s3cli       s3iface.S3API
+		queryParser QueryParser
+	}
+
+	visibilityRecord archiver.ArchiveVisibilityRequest
+
+	queryVisibilityRequest struct {
+		domainID      string
+		pageSize      int
+		nextPageToken []byte
+		parsedQuery   *parsedQuery
+	}
+)
+
+const (
+	errEncodeVisibilityRecord = "failed to encode visibility record"
+	indexKeyStartTimeout      = "startTimeout"
+	indexKeyCloseTimeout      = "closeTimeout"
+)
+
+// NewVisibilityArchiver creates a new archiver.VisibilityArchiver based on s3
+func NewVisibilityArchiver(
+	container *archiver.VisibilityBootstrapContainer,
+	config *config.S3Archiver,
+) (archiver.VisibilityArchiver, error) {
+	return newVisibilityArchiver(container, config)
+}
+
+func newVisibilityArchiver(
+	container *archiver.VisibilityBootstrapContainer,
+	config *config.S3Archiver) (*visibilityArchiver, error) {
+	s3Config := &aws.Config{
+		Endpoint:         config.Endpoint,
+		Region:           aws.String(config.Region),
+		S3ForcePathStyle: aws.Bool(config.S3ForcePathStyle),
+		MaxRetries:       aws.Int(0),
+	}
+	sess, err := session.NewSession(s3Config)
+	if err != nil {
+		return nil, err
+	}
+	return &visibilityArchiver{
+		container:   container,
+		s3cli:       s3.New(sess),
+		queryParser: NewQueryParser(),
+	}, nil
+}
+
+func (v *visibilityArchiver) Archive(
+	ctx context.Context,
+	URI archiver.URI,
+	request *archiver.ArchiveVisibilityRequest,
+	opts ...archiver.ArchiveOption,
+) (err error) {
+
+	featureCatalog := archiver.GetFeatureCatalog(opts...)
+	defer func() {
+		if err != nil {
+			if featureCatalog.NonRetriableError != nil && !isRetryableError(err) {
+				err = featureCatalog.NonRetriableError()
+			}
+		}
+	}()
+
+	logger := archiver.TagLoggerWithArchiveVisibilityRequestAndURI(v.container.Logger, request, URI.String())
+
+	if err := v.ValidateURI(URI); err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(archiver.ErrReasonInvalidURI), tag.Error(err))
+		return err
+	}
+
+	if err := archiver.ValidateVisibilityArchivalRequest(request); err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(archiver.ErrReasonInvalidArchiveRequest), tag.Error(err))
+		return err
+	}
+
+	encodedVisibilityRecord, err := encode(request)
+	if err != nil {
+		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeVisibilityRecord), tag.Error(err))
+		return err
+	}
+
+	// Upload archive to all indexes
```

Comment:
Probably worth explaining how data is stored and indexed in README.

Reply:
Done!